kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/compose/compose.kt

```diff
@@ -0,0 +1,56 @@
+package io.kotest.matchers.compose
+
+import io.kotest.assertions.fail
+import io.kotest.matchers.Matcher
+import io.kotest.matchers.MatcherResult
+import kotlin.reflect.KProperty1
+
+/**
+ * Compose matchers along with the property to extract to test against.
+ * All matchers must be the same type `T`. If any matcher fails, composed matcher fails, too.
+ */
+@Suppress("UNCHECKED_CAST")
+fun <T : Any?> Matcher.Companion.compose(
+   vararg pairs: Pair<Matcher<*>, KProperty1<T, *>>
+): Matcher<T> = object : Matcher<T> {
+   override fun test(value: T): MatcherResult {
+      val results = pairs.mapIndexed {i , (matcher, prop) ->
+         runCatching {
+            (matcher as Matcher<Any?>).test(prop(value))
+         }.onFailure {
+            fail("Mismatching types in argument ${i + 1} for composed matcher:\n\n${it.message}")
+         }.getOrThrow()
+      }
+
+      return MatcherResult(
+         results.all { it.passed() },
+         { results.joinToString(separator = "\n") { it.failureMessage() } },
+         { results.joinToString(separator = "\n") { it.negatedFailureMessage() } },
+      )
+   }
+}
+
+/**
+ * Compose matchers. All matchers must be the same type `T`. If any matcher fails, composed matcher fails, too.
+ */
+@Suppress("UNCHECKED_CAST")
+fun <T : Any?> Matcher.Companion.compose(
+   vararg matchers: Matcher<T>
+): Matcher<T> = object : Matcher<T> {
+   override fun test(value: T): MatcherResult {
+      val results = matchers.mapIndexed { i, matcher  ->
+         runCatching {
+            (matcher as Matcher<Any?>).test(value)
+         }.onFailure {
+            fail("Mismatching types in argument ${i + 1} for composed matcher:\n\n${it.message}")
+         }.getOrThrow()
+      }
```

Comment:
```suggestion
      val results = matchers.map { matcher -> matcher.test(value) }
```

Can be simplified, since we have type constraints on both `value` and `matcher`. Trying to compose matchers for different types gives a warning: (Will fail completely from next Kotlin version)
<img width="1136" alt="image" src="https://user-images.githubusercontent.com/316857/221016382-536a4739-92dd-4279-bfa5-4e3629dcfac8.png">

If you explicitly state the type of the matcher, it will fail compilation: 
<img width="596" alt="image" src="https://user-images.githubusercontent.com/316857/221016484-27a772d1-622d-477c-aecd-4117f5b753a9.png">